builtin/providers/aws/resource_aws_s3_bucket_notification.go

```diff
@@ -0,0 +1,453 @@
+package aws
+
+import (
+	"bytes"
+	"fmt"
+	"log"
+	"strings"
+
+	"github.com/hashicorp/terraform/helper/hashcode"
+	"github.com/hashicorp/terraform/helper/schema"
+
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/aws/awserr"
+	"github.com/aws/aws-sdk-go/service/s3"
+)
+
+func resourceAwsS3BucketNotification() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceAwsS3BucketNotificationPut,
+		Read:   resourceAwsS3BucketNotificationRead,
+		Update: resourceAwsS3BucketNotificationPut,
+		Delete: resourceAwsS3BucketNotificationDelete,
+
+		Schema: map[string]*schema.Schema{
+			"bucket": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: true,
+			},
+
+			"topic": &schema.Schema{
+				Type:     schema.TypeSet,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"id": &schema.Schema{
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"filter_prefix": &schema.Schema{
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"filter_suffix": &schema.Schema{
+							Type:     schema.TypeString,
+							Optional: true,
+						},
+						"topic": &schema.Schema{
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"events": &schema.Schema{
+							Type:     schema.TypeSet,
+							Required: true,
+							Elem:     &schema.Schema{Type: schema.TypeString},
+							Set:      schema.HashString,
+						},
+					},
+				},
+				Set: func(v interface{}) int {
+					var buf bytes.Buffer
+					m := v.(map[string]interface{})
+					buf.WriteString(fmt.Sprintf("%t-", m["filter_prefix"].(string)))
+					buf.WriteString(fmt.Sprintf("%t-", m["filter_suffix"].(string)))
+					return hashcode.String(buf.String())
```

Comment:
Poking at this further, it seems that if omitted, the `id` will get populated by the API. This is troublesome because we'll then see a diff in future plans if we don't have that populated.